### ISSUE TITLE
Bug: Keyvaluemaps goal and update build option deletes existing KVM in versions >2.4.0

### ISSUE THREAD

**Issue by michael-kirkegaard-solita** [apigee/apigee-config-maven-plugin#192]
In the versions greater than 2.4.0 the keyvaluemaps goal for existing KVMs is incorrectly deleting and recreating the KVM. Expected behavior is to only update, not delete and re-create.

In version 1.x.x, we see only updates as expected
![image](https://github.com/apigee/apigee-config-maven-plugin/assets/119622241/241cdcd2-f0d1-40c0-80f6-a88e0cff0cce)

In version 2.x.x the sync and and the update options are equivalent, but the update should NOT delete and re-create
![image](https://github.com/apigee/apigee-config-maven-plugin/assets/119622241/df398b76-f482-4a7a-868e-0f1a3b0ead82)

**Comment by ssvaidyanathan** [comment]
Ack @michael-kirkegaard-solita 

Initially, the APIs for managing the entries were not available, so the plugin was only supporting creation and deletion of the KVM itself. Now they are available, so let me do some analysis and find if it can be similar to what was available in 1.x.x

Will get back to you

**Comment by ssvaidyanathan** [comment]
@michael-kirkegaard-solita - Did some analysis, in Apigee X/hybrid we do not have an API similar to [this](https://apidocs.apigee.com/docs/key-value-maps/1/routes/organizations/%7Borg_name%7D/environments/%7Benv_name%7D/keyvaluemaps/%7Bmap_name%7D/entries/%7Bentry_name%7D/post) one available in Edge. 

So in Edge, using 1.x, we were able to update the entries directly. But the same does not exist in X, only operations available were to individually create and update the entries. 

I could make the plugin use that to delete and create each entry from the json you provide. Do you think that's better than deleting and creating the KVM? The issue with this approach is that, lets say you had 3 entries in your json file and you run the command, it will create those entries for you. Any updates you make to existing entries will also get updated as each of those entries are deleted and recreated. Now lets say you want to delete one of the three entries and you remove that from your json. The plugin will not be able to delete that from Apigee as that is not available in the source json itself. The only option for you in that case is to run the plugin using the "sync" option which will delete the entire KVM and create the KVM again with the two entries you have now in the json.

And that is exactly what the current feature is providing for update. 

Let me know if your thoughts. My preference is to leave the way it is as it will be more consistent and will be in sync with what you have in your json file.

**Comment by michael-kirkegaard-solita** [comment]
@ssvaidyanathan Since the Apigee X API does not support upsert via POST method, deleting and creating individual entries would be the closest to an "upsert" operation we could get, that would support the expected behavior. 

The description of the `update` build options is the following:
`update - Update when found; create when not found, updates individual entries for kvms. Refreshes all config to reflect edge.json.`
Whereas the `sync` description is;
`sync - Delete and recreate`
So the behavior you want to maintain in the `update` build-option, is actually what `sync` is documented to do.

As you describe it yourself - you are trying to sync the json-file and the deployed configurations. So that should be performed with a "sync" build-option. The `sync` build-option is documented as a create-and-delete option. This true for both 1.x and 2.x.

However, in 1.x the `update` build options is different from the `sync` option, in that it does not delete any entries, It only updates and overwrites exactly what is defined in json-file.
In 2,.x `update` and `sync` is performing the exact same steps, although they are still documented differently. I would expect the two build-options to have difference behaviors, because If the `update` and the `sync` build-options are doing the same thing, why even have two different ones?

**Comment by michael-kirkegaard-solita** [comment]
@ssvaidyanathan Furthermore for my use-case, the problem with deleting the entire KVM is that I lose the ability to have a distributed setup. Lets say you have 2 different pom-projects, A and B, each with their own edge.json file (they are stored in different git-repos). Both projects uses the same KVM; Deployment A defines some of the values ("a.1", "a.2", "a.3"), Deployment B defines some the values ("b,1", "b.2", "b.3").
In 1.x - both of these can deploy independently of each other and all entries will be guaranteed to exists in the KVM.
in 2.x there is no option to do this, because `update` is now deleting other entries, so I will be forced to combine projects into a single central deployment, instead of a distributed deployment.

It is still possible to achieve what you described above by using the appropriately name `sync` option, but in 2x it no longer possible to achieve my use-case. My organization has just migrated to Apigee X, and since we have 200+ distributed deployments, we will be forced to find another ci/cd tool for our pipelines, since centralizing all edge.json files is not an option.

**Comment by michael-kirkegaard-solita** [comment]
@ssvaidyanathan for using the new Apigee X/Hybrid APIs to replace the method you linked above, I propose that the `update` logic should be similar to the pseudocode below:

```
for each json-kvm in edge.json {
    // create if not found
    if (json-kvm does not exist in apigee) {
        POST kvm to apigee
    }
    // updates individual entries for kvms
    for each entry in json-kvm {
        if (entry exist in apigee-kvm) {
            DELETE entry from apigee-kvm
        }
        POST entry to apigee-kvm
    }
}
```

This would guarantee the all apigee config is refreshed to reflect edge.json. But elements not defined in the edge.json is unaltered.

**Comment by ssvaidyanathan** [comment]
Thanks for the great discussion. This is exactly what I wanted to know. As i said in my earlier comment, the update option of KVM is like sync due to the unavailability of the POST /entries API that can update existing KVM entries. 

Are you ok if you still have an entry in the KVM but not in your actual json ? Lets say you had 3 entries and now you removed one from your json. The plugin might not be able to pick that as its using json file as the source. So the 3rd entry will still be in Apigee's KVM. Are you ok with that?

**Comment by michael-kirkegaard-solita** [comment]
Yes, that is acceptable.

The `update` build-option can then be considered a guarantee that the edge.json config exists and will not impact any other config.

And if you want to guarantee that no other config exists except what is in the edge.json, then the `sync` build option will still do that.

**Comment by ssvaidyanathan** [comment]
@michael-kirkegaard-solita - let me work on that change (I will need to add this logic to all the different KVM scopes)
Let me release a rc version of the plugin with the change. Would appreciate if you can run some tests for me and confirm if its working as expected.

**Comment by ssvaidyanathan** [comment]
HI @michael-kirkegaard-solita - Apologies

I was out on vacation. Will work on it this week.

**Comment by ssvaidyanathan** [comment]
@michael-kirkegaard-solita - I just released a "rc" with the fix (version: `v2.6.1-rc1`)
Please update your pom file to point to that version and test it out. Let me know if it works as we discussed.

Appreciate if you can test the other scopes of KVM too using the plugin - "proxy" and "org" 
Also try various scenarios with the config options - create, update, delete, sync

Once you confirm, I will release it as `v2.7.0`

**Comment by ssvaidyanathan** [comment]
@michael-kirkegaard-solita - were you able to test?

**Comment by michael-kirkegaard-solita** [comment]
Hi @ssvaidyanathan,

I have finished my testing, and concluded the following:
The `update` build option is now behaving as we agreed on prior in the discussion.

Furthermore, I have compared the keyvaluemaps goal in version 2.4.0 to 2.6.1-rc1 and determined that the other build options are also behaving as expected in both versions.

I think these changes are ready for the 2.7.0 release.

**Comment by michael-kirkegaard-solita** [comment]
I see that forward-slash encoding is a problem although it is allowed in the body, so here is a simpler example for the entry-names issue with special characters from above:

![image](https://github.com/apigee/apigee-config-maven-plugin/assets/119622241/50f7c379-fa96-464f-b139-399e66b5957a)

**Comment by ssvaidyanathan** [comment]
Thanks a lot @michael-kirkegaard-solita 
Let me fix those and release a rc2 so that we test all those scenarios as well.
Am glad you tested these cases. I clearly had not. Was focussing on getting the flow we discussed to work :)
will provide an update as soon as I have the rc2 pushed

**Comment by ssvaidyanathan** [comment]
@michael-kirkegaard-solita - quick update

I have the fix for characters like `#?+` but the API is failing if the entry includes a `/`. I am working with the team to see what can be done.

**Comment by ssvaidyanathan** [comment]
@michael-kirkegaard-solita - looks like the Google Front End (thats in front of the Apigee API) is not allowing the "/" even if they are URL encoded. So am thinking of implementing a check to see if there are any "/" in the key names and throw an error while creating or updating.

**Comment by ssvaidyanathan** [comment]
@michael-kirkegaard-solita - Just released `v2.6.1-rc2`, can you please try out some test cases

**Comment by ssvaidyanathan** [comment]
@michael-kirkegaard-solita - were you able to test?

**Comment by michael-kirkegaard-solita** [comment]
Hi @ssvaidyanathan 
Sorry about the delay, thanks for the patience.

Yes I was able to test the release candidate. The functionality seems to work as intended using "update" build option
- KVMs are created if they don't exist
- KVM entries are deployed
- Other KVM entries are not deleted if the KVM existed before the update
- Invalid keys (containing "/") are correctly blocked.

Overall, very satisfied with the results.

**Comment by ssvaidyanathan** [comment]
@michael-kirkegaard-solita - thanks for confirming

Not sure you saw the latest Apigee release note - Update for KVM entry is now available as an Apigee API :)

So I went ahead and released v2.7.0 just now with the update command (instead of the delete/create calls). Ran some test and they all seem to work as expected. Appreciate if you can try the new version and let me know if there are any issues. Thanks again for all the collaboration
